src/SystemState.ts

```diff
@@ -101,4 +106,45 @@ export class SystemState {
         checkIntegerInRange("Group index", iGroup, 0, this._nGroups - 1);
         checkIntegerInRange("Particle index", iParticle, 0, this._nParticles - 1);
     }
+
+    /**
+     * Apply a reordering to the particles in every group according to a reordering parameter which defines
+     * a new order for each group.
+     * @param reordering {@link https://github.com/scijs/ndarray | NdArray } whose first dimension defines group index
+     * and whose second dimension defines a new order of particles for each group. For example, if this state has three
+     * particles per group, a reordering of `[2, 0, 1]` for any group would imply that its current third particle should
+     * be reordered to the first position, its current first particle to the second position, and its current second
+     * particle to the third position.
+     */
+    public reorder(reordering: ndarray.NdArray) {
+        const shape = reordering.shape;
+        if (shape.length !== 2 || shape[0] !== this._nGroups || shape[1] !== this._nParticles) {
+            throw new DustParameterError(
+                "Unexpected reordering shape. " +
+                    `Expected [${this._nGroups},${this._nParticles}] but got ${JSON.stringify(shape)}.`
+            );
+        }
+        // Check that each reordering contains expected values
+        for (let iGroup = 0; iGroup < this._nGroups; iGroup++) {
+            const reorder = particleStateToArray(reordering.pick(iGroup, null));
+            const sortedReorder = reorder.toSorted();
+            // Each reordering should contain every index value once so, when sorted, each value should equal its index
+            if (sortedReorder.some((value: number, index: number) => value !== index)) {
+                throw new DustParameterError(`Unexpected reorder values: ${JSON.stringify(reorder)}`);
+            }
+        }
+
+        const reordered = this.newState();
+        for (let iGroup = 0; iGroup < this._nGroups; iGroup++) {
```

Comment:
We might want a check-less version for use within the filter where we generate our own index - consider splitting this into two (`reorder` and `reorderNoCheck` or something?)

Reply:
I've just split that out into a private method with a comment for now, we might provide another means of access when we do the particle filter.